kernel/mutex.c

```diff
@@ -200,18 +182,32 @@ Z_SYSCALL_HANDLER(k_mutex_lock, mutex, timeout)
 }
 #endif
 
-void z_impl_k_mutex_unlock(struct k_mutex *mutex)
+int z_impl_k_mutex_unlock(struct k_mutex *mutex)
 {
 	struct k_thread *new_owner;
 
-	__ASSERT(mutex->lock_count > 0U, "");
-	__ASSERT(mutex->owner == _current, "");
+	/*
+	 * If a thread attempts to unlock a mutex that it has not locked or a
+	 * mutex which is unlocked, undefined behavior results.
+	 *
+	 * Similar to posix, if a thread attempts to unlock a mutex which is
+	 * unlocked, k_mutex_unlock() shall behave as "Unlock When Not Owner"
+	 * below.
+	 */
+	if (mutex->lock_count == 0U) {
+		return -EPERM;
```

Comment:
Please use the same error codes as sys_mutex(), for consistency.
This should be -EINVAL